src/swish/erlang.ss

```diff
@@ -943,140 +945,255 @@
+           (eq? (datum guard) 'guard)
+           (test g #,(f #'more) #,fail)]
+          [((sub-match v pattern))
+           (and object? (eq? (datum sub-match) 'sub-match))
+           ;; sub-match pattern does not allow guard
+           (convert #`(v pattern #,fail #,body))]
+          [((handle-fields v field-spec ...))
+           (and object? (eq? (datum handle-fields) 'handle-fields))
+           (let ()
+             (define (do-field field dest-var options body)
+               ;; We don't check whether handler emits a bind for dest-var.
+               ;; If the code is in a library, the compiler will complain for us.
+               (generate context #f #f fail body
+                 (or (handle-field #'v field dest-var options context)
+                     (bad-syntax x "unknown field" field))))
+             (let match-extended-help ([specs #'(field-spec ...)])
+               (syntax-case specs (unquote unquote-splicing)
+                 [() body]
+                 [((unquote field) . rest)
+                  (if (identifier? #'field)
+                      (do-field #'field #'field '()
+                        (match-extended-help #'rest))
+                      (bad-pattern x))]
+                 [((unquote-splicing var) . rest)
+                  (if (identifier? #'var)
+                      (with-temporaries (tmp)
+                        (do-field #'var #'tmp '()
+                          (test (equal? tmp var)
+                            #,(match-extended-help #'rest)
+                            #,fail)))
+                      (bad-pattern x))]
+                 [([field pattern option ...] . rest)
+                  (with-temporaries (tmp)
+                    (do-field #'field #'tmp #'(option ...)
+                      (convert #`(tmp pattern #,fail #,(match-extended-help #'rest)))))]
+                 [other (bad-pattern x)])))]
+          [oops (syntax-error context "bad output from handle-object for")])))
+    (define (convert x)
+      (syntax-case x (unquote unquote-splicing quasiquote)
+        [(e (unquote (v <= pattern)) fail body)
+         (and (identifier? #'v) (eq? (datum <=) '<=))
+         (bind-var v e (convert #'(v pattern fail body)))]
+        [(e (unquote v) fail body)
+         (cond
+          [(eq? (datum v) '_) #'body]
+          [(identifier? #'v) (bind-var v e #'body)]
+          [else (bad-pattern x)])]
+        [(e (unquote-splicing var) fail body)
+         (if (identifier? #'var)
+             (test (equal? e var) body fail)
+             (bad-pattern x))]
+        [(e (quasiquote (type spec ...)) fail body)
+         (call-with-values
+           (lambda () (get-extended lookup #'type))
+           (case-lambda
+            [(fail)
+             (cond
+              [(eq? fail 'bad-pattern) (bad-pattern x)]
+              [else (bad-syntax x "unknown type" #'type)])]
+            [(handle-object handle-field)
+             (let ([pattern #'(quasiquote (type spec ...))])
```

Comment:
Rebuilding the ``#`(quasiquote (type spec ...))`` here was a mistake: if `handle-object` or `generate` complain about the pattern, we'll get the wrong source information in the error message. If we extract the pattern subform from the input, we'll get the right source.

Reply:
Fixed.